java/arcs/sdk/android/storage/service/StorageService.kt

```diff
@@ -69,6 +69,10 @@ open class StorageService : ResurrectorService() {
             Thread(it).apply { name = "WriteBack #$id" }
         }.asCoroutineDispatcher() + SupervisorJob()
     )
+    protected open val cleanupTaskClass = PeriodicCleanupTask::class.java
```

Comment:
I know we've seen init order issues with open vals in some cases. Two other optionsto consider:
1. add these to the config (see line 78)
2. add them to the constructor as params

I think open val with subclass override is generally fragile and not a preferred pattern, though we had to do it due to scope of otherwise required refactoring in past? @jasonwyatt for Kt style guidance.

Reply:
I remembered that the crashes are similar to the one discussed in this article: https://medium.com/keepsafe-engineering/an-in-depth-look-at-kotlins-initializers-a0420fcbf546

For 1, config itself is an open val so it has similar risks. But I think moving these two params to config makes it more clear as the config class includes the ttl intervals as well.
For 2, Android services are constructed by system so we can't pass params.

Reply:
why do you need them open? can you add more context to the pr description?

Reply:
Updated.

Reply:
To add some context: I want to add a customized Worker which does the cleanup as well as logging. The Worker also needs to be Dagger injectable to get the customized logger in google3 so I can't just update the Worker code in GitHub.

Reply:
Thanks, as suggested by Walter, move it to the config object?

Reply:
We can avoid the issues with base class initializers seeing the wrong thing by making sure  to override *methods*, not *fields*.

We could do this with `vals` by overriding getters rather than the actual field value, i.e.,
`open fun cleanupTaskClass() = PeriodicCleanupTask::class.java`


```
override val cleanupTaskClass: Class<Worker>
    get() = PeriodicCleanupTask::class.java
```

However, this is a bit fragile as the subclasser needs to remember this detail. We can instead just move to defining open methods, instead:

Reply:
(And I do like the idea of putting everything into a single `Config` data class and just defining one open method to provide the config)

Reply:
Agreed. This open val is some Kotlin design flaw IMO as it's hard to use and you simply can't describe the right way to use it in less than 100 words. So let's just keep these newly added stuff in a single config object so we can fix all the issues in this one place.

Reply:
What do you think of this pattern? It prevents the ability to incorrectly override `val`, and only creates the item once:

Base class:
```
val config: Config = config()
open fun provideConfig() = DefaultConfig()
```

Subclass:
```
override fun provideConfig() = CustomConfig()
```

Reply:
How does this avoid the issue?

Reply:
The `val config` itself is not open at all, avoiding the potential to incorrectly override it.

It  will be initialized with the result of executing the `provideThing` method. Even in the base class initializer, the subclass `provideConfig` will be executed, so you'll get the value provided by the subclass `provideConfig`, if there is one.

Reply:
I'm neutral on this one. I think overriding vals still provide more context for readability as otherwise you need to explain in comments why you are doing this instead of overriding vals.

Reply:
If we allow overriding vals we need to explain in comments that it's very careful to avoid using that val during any base class initialization steps. So we need to take care to never reference the field during base class initialization (including initializing any fields in the base class), OR any initializer using the val needs to be wrapped in a `lazy` so that it doesn't attempt to access the field directly during object initialization.

Either way, there's documentation to be done.

Reply:
I think we could enforce lazy init by declaring the field with Lazy<T>, e.g.

```
open val foo: Lazy<Foo> = lazy { value }
```

This means all overriders also have to be lazy.

Reply:
Can we move the refactoring/pattern change to a different PR? I need this PR for logging stuff. A separate PR would also help future references on the new patterns.

Reply:
@cromwellian I don't think that will help here. The issue is with usages during base class initialization. Even if the subclassed val is lazy, it simply doesn't exist yet during base class initialization time.

To avoid the problem when using open val, we need to take care that the base class doesn't ever access the field until after initialization expecting to get a subclass value.

Reply:
@yuangu-google oh yeah, for sure.

Reply:
Jibble, you're right, it fixes the NPE issue, but ends up with the base class getting a stale value before the subclass initializer sees a different value. 

I think JetBrains should have some kind of way of declaring that an open val can't be referenced in initializers. To me, any properly type-checked program should not through an NPE. 

I mean, the easiest thing to do is just use an open fun, not a val field.  In Java, you'd be forced to do this anyway. I guess  this is the one thing Java has over Kotlin. :)

Reply:
Totally agreed. This is a Kotlin design issue.

Reply:
Are things really that much better in Java? I remember getting tripped up by this in my early days when using Java...You still have to remember that if you declare a field with the same name in a derived class, it's not available during construction of the base class; you'll get the value of the base class's field instead.

I found that Android Studio does have an Inspection (currently turned off for me) called "Instance field may not be initialized".  Enabling it does seem to flag any issues for me. It looks like it's off by default because it uses a conservative dataflow algorithm and may report false issues.

Reply:
I don't have any valid suggestion, but just noting that the this is further complicated by the fact that the AiAiStorageService subclass needs to read device config to set this config, which is available only after onCreate.